File: gherkin/python/gherkin/pickles/compiler.py
import re

from ..count_symbols import count_symbols
from ..stream.id_generator import IdGenerator


class Compiler(object):
    def __init__(self, id_generator=None):
        self.id_generator = id_generator
        if self.id_generator is None:
            self.id_generator = IdGenerator()

    def compile(self, gherkin_document):
        pickles = []
        if 'feature' not in gherkin_document:
            return pickles

        feature = gherkin_document['feature']
        if not feature['children']:
            return pickles

        uri = gherkin_document['uri']
        feature_tags = feature['tags']
        language = feature['language']
        background_steps = []
        for scenario_definition in feature['children']:
            if 'background' in scenario_definition:
                if scenario_definition['background']['steps']:
                    background_steps += scenario_definition['background']['steps']
            elif 'rule' in scenario_definition:
                self._compile_rule(uri, feature_tags, background_steps, scenario_definition['rule'], language, pickles)
            else:
                scenario = scenario_definition['scenario']
                args = (uri, feature_tags, background_steps, scenario, language, pickles)
                if not scenario['examples']:
                    self._compile_scenario(*args)
                else:
                    self._compile_scenario_outline(*args)
        return pickles

    def _compile_rule(self, uri, feature_tags, feature_background_steps, rule, language, pickles):
        tags = list(feature_tags) + list(rule['tags'])
        background_steps = []
        background_steps += feature_background_steps
        for scenario_definition in rule['children']:
            if 'background' in scenario_definition:
                if scenario_definition['background']['steps']:
                    background_steps += scenario_definition['background']['steps']
            else:
                scenario = scenario_definition['scenario']
                args = (uri, tags, background_steps, scenario, language, pickles)
                if not scenario['examples']:
                    self._compile_scenario(*args)
                else:
                    self._compile_scenario_outline(*args)
        return pickles

    def _compile_scenario(self, uri, inherited_tags, background_steps, scenario, language, pickles):
        tags = list(inherited_tags) + list(scenario['tags'])
        last_keyword_type = 'Unknown'
        steps = list()
        if scenario['steps']:
            for step in background_steps + scenario['steps']:
                last_keyword_type = last_keyword_type if step['keywordType'] == 'Conjunction' else step['keywordType']
                steps.append(self._pickle_step(step, last_keyword_type))

        pickle = {
            'astNodeIds': [scenario['id']],
            'id': self.id_generator.get_next_id(),
            'tags': self._pickle_tags(tags),
            'name': scenario['name'],
            'language': language,
            'steps': steps,
            'uri': uri
        }
        pickles.append(pickle)


    def _compile_scenario_outline(self, uri, inherited_tags, background_steps, scenario, language, pickles):
        for examples in (e for e in scenario['examples'] if 'tableHeader' in e):
            variable_cells = examples['tableHeader']['cells']

            for values in examples['tableBody']:
                value_cells = values['cells']
                tags = list(inherited_tags) + list(scenario['tags']) + list(examples['tags'])
                last_keyword_type = None
                steps = list()
                if scenario['steps']:
                    for step in background_steps:
                        last_keyword_type = last_keyword_type if step['keywordType'] == 'Conjunction' else step['keywordType']
                        steps.append(self._pickle_step(step, last_keyword_type))

                if scenario['steps']:
                    for outline_step in scenario['steps']:
                        last_keyword_type = last_keyword_type if outline_step['keywordType'] == 'Conjunction' else outline_step['keywordType']
                        step_text = self._interpolate(
                            outline_step['text'],
                            variable_cells,
                            value_cells)
                        argument = self._create_pickle_arguments(
                            outline_step,
                            variable_cells,
                            value_cells)
                        _pickle_step = {
                            'astNodeIds': [outline_step['id'], values['id']],
                            'id': self.id_generator.get_next_id(),
                            'type': last_keyword_type,
                            'text': step_text
                        }
                        if argument is not None:
                            _pickle_step['argument'] = argument

                        steps.append(_pickle_step)

                pickle = {
                    'astNodeIds': [scenario['id'], values['id']],
                    'id': self.id_generator.get_next_id(),
                    'name': self._interpolate(
                        scenario['name'],
                        variable_cells,
                        value_cells),
                    'language': language,
                    'steps': steps,
                    'tags': self._pickle_tags(tags),
                    'uri': uri
                }
                pickles.append(pickle)

    def _create_pickle_arguments(self, step, variables, values):
        if 'dataTable' in step:
            table = {'rows': []}
            for row in step['dataTable']['rows']:
                cells = [
                    {
                        'value': self._interpolate(cell['value'], variables, values)
                    } for cell in row['cells']
                ]
                table['rows'].append({'cells': cells})
            return {'dataTable': table}

        elif 'docString' in step:
            argument = step['docString']
            docstring = {
                'content': self._interpolate(argument['content'], variables, values)
            }
            if 'mediaType' in argument:
                docstring['mediaType'] = self._interpolate(argument['mediaType'], variables, values)
            return {'docString': docstring}
        else:
            return None

    def _interpolate(self, name, variable_cells, value_cells):
        if name is None:
            return name

        for n, variable_cell in enumerate(variable_cells):
            value_cell = value_cells[n]
            # For the case of trailing backslash, re-escaping backslashes are needed
            reescaped_value = re.sub(r'\\', r'\\\\', value_cell['value'])
            name = re.sub(
                u'<{0[value]}>'.format(variable_cell),
                reescaped_value,
                name
            )
        return name

    def _pickle_step(self, step, keyword_type):
        pickle_step = {
            'astNodeIds': [step['id']],
            'id': self.id_generator.get_next_id(),
            'type': keyword_type,
            'text': step['text'],
        }
        argument = self._create_pickle_arguments(
            step,
            [],
            [])
        if argument is not None:
            pickle_step['argument'] = argument
        return pickle_step


    def _pickle_tags(self, tags):
        return [self._pickle_tag(tag) for tag in tags]


    def _pickle_tag(self, tag):
        return {
            'astNodeId': tag['id'],
            'name': tag['name']
        }

File: gherkin/javascript/lib/gherkin/pickles/compiler.js
var countSymbols = require('../count_symbols');

function Compiler() {
  this.compile = function (gherkin_document) {
    var pickles = [];

    if (gherkin_document.feature == null) return pickles;

    var feature = gherkin_document.feature;
    var language = feature.language;
    var tags = feature.tags;
    var backgroundSteps = [];

    build(pickles, language, tags, backgroundSteps, feature);
    return pickles;
  };

  function build(pickles, language, tags, parentBackgroundSteps, parent) {
    var backgroundSteps = parentBackgroundSteps.slice(0); // dup
    parent.children.forEach(function (child) {
      if (child.type === 'Background') {
        backgroundSteps = backgroundSteps.concat(pickleSteps(child));
      } else if (child.type === 'Rule') {
        build(pickles, language, tags, backgroundSteps, child)
      } else {
        var scenario = child
        if (scenario.examples.length === 0) {
          compileScenario(tags, backgroundSteps, scenario, language, pickles);
        } else {
          compileScenarioOutline(tags, backgroundSteps, scenario, language, pickles);
        }
      }
    });
  }

  function compileScenario(featureTags, backgroundSteps, scenario, language, pickles) {
    var steps = scenario.steps.length == 0 ? [] : [].concat(backgroundSteps);

    var tags = [].concat(featureTags).concat(scenario.tags);

    scenario.steps.forEach(function (step) {
      steps.push(pickleStep(step));
    });

    var pickle = {
      tags: pickleTags(tags),
      name: scenario.name,
      language: language,
      locations: [pickleLocation(scenario.location)],
      steps: steps
    };
    pickles.push(pickle);
  }

  function compileScenarioOutline(featureTags, backgroundSteps, scenario, language, pickles) {
    scenario.examples.filter(function(e) { return e.tableHeader != undefined; }).forEach(function (examples) {
      var variableCells = examples.tableHeader.cells;
      examples.tableBody.forEach(function (values) {
        var valueCells = values.cells;
        var steps = scenario.steps.length == 0 ? [] : [].concat(backgroundSteps);
        var tags = [].concat(featureTags).concat(scenario.tags).concat(examples.tags);

        scenario.steps.forEach(function (scenarioOutlineStep) {
          var stepText = interpolate(scenarioOutlineStep.text, variableCells, valueCells);
          var args = createPickleArguments(scenarioOutlineStep.argument, variableCells, valueCells);
          var pickleStep = {
            text: stepText,
            arguments: args,
            locations: [
              pickleLocation(values.location),
              pickleStepLocation(scenarioOutlineStep)
            ]
          };
          steps.push(pickleStep);
        });

        var pickle = {
          name: interpolate(scenario.name, variableCells, valueCells),
          language: language,
          steps: steps,
          tags: pickleTags(tags),
          locations: [
            pickleLocation(values.location),
            pickleLocation(scenario.location)
          ]
        };
        pickles.push(pickle);

      });
    });
  }

  function createPickleArguments(argument, variableCells, valueCells) {
    var result = [];
    if (!argument) return result;
    if (argument.type === 'DataTable') {
      var table = {
        rows: argument.rows.map(function (row) {
          return {
            cells: row.cells.map(function (cell) {
              return {
                location: pickleLocation(cell.location),
                value: interpolate(cell.value, variableCells, valueCells)
              };
            })
          };
        })
      };
      result.push(table);
    } else if (argument.type === 'DocString') {
      var docString = {
        location: pickleLocation(argument.location),
        content: interpolate(argument.content, variableCells, valueCells),
      };
      if(argument.contentType) {
        docString.contentType = interpolate(argument.contentType, variableCells, valueCells);
      }
      result.push(docString);
    } else {
      throw Error('Internal error');
    }
    return result;
  }

  function interpolate(name, variableCells, valueCells) {
    variableCells.forEach(function (variableCell, n) {
      var valueCell = valueCells[n];
      var search = new RegExp('<' + variableCell.value + '>', 'g');
      // JS Specific - dollar sign needs to be escaped with another dollar sign
      // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#Specifying_a_string_as_a_parameter
      var replacement = valueCell.value.replace(new RegExp('\\$', 'g'), '$$$$')
      name = name.replace(search, replacement);
    });
    return name;
  }

  function pickleSteps(scenarioDefinition) {
    return scenarioDefinition.steps.map(function (step) {
      return pickleStep(step);
    });
  }

  function pickleStep(step) {
    return {
      text: step.text,
      arguments: createPickleArguments(step.argument, [], []),
      locations: [pickleStepLocation(step)]
    }
  }

  function pickleStepLocation(step) {
    return {
      line: step.location.line,
      column: step.location.column + (step.keyword ? countSymbols(step.keyword) : 0)
    };
  }

  function pickleLocation(location) {
    return {
      line: location.line,
      column: location.column
    }
  }

  function pickleTags(tags) {
    return tags.map(function (tag) {
      return pickleTag(tag);
    });
  }

  function pickleTag(tag) {
    return {
      name: tag.name,
      location: pickleLocation(tag.location)
    };
  }
}

module.exports = Compiler;
